[DataViews]: expose self-sufficient subcomponents (Search, Filters, ViewConfig, Layout, Footer)

retrofox opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Part of WOOA7S-113

## Proposed Changes

* Expose `Search`, `ViewConfig`, `Layout`, `Pagination`, `LayoutSwitcher`, and `BulkActionToolbar` as self-contained subcomponents of DataViews.
* Update the Free Composition story to use the new subcomponents and demonstrate context-based rendering.

## Why are these changes being made?

We want to enable developers to compose their own custom DataViews-based UIs using smaller, self-contained parts. These subcomponents are already used internally by the main `DataViews` layout and are capable of rendering in isolation when context is properly provided.

This improves flexibility in layout and composition while avoiding the need to duplicate internal logic outside the package.

## Testing Instructions

* Run the Storybook with:

```
yarn dataviews:storybook:start
```

* Open the `FreeComposition` story.
* Verify that all subcomponents (`Search`, `ViewConfig`, `Filters`, `Layout`, and `Pagination`) are rendered as expected and interact correctly with the context.
* Confirm that custom styling via `className` works on `<DataViewsFooter />`.

https://github.com/user-attachments/assets/ec6d1697-48d2-4071-8b17-92a52299373e

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

retrofox committed:
Export DataViews subcomponents as named exports

Improve component discoverability and tree-shaking capabilities

retrofox committed:
iterate over Free Composition story

retrofox committed:
changelog

retrofox commented on packages/dataviews/src/components/dataviews-footer/index.tsx:
```diff
@@ -17,7 +17,11 @@ import { LAYOUT_GRID, LAYOUT_TABLE } from '../../constants';
 
-export default function DataViewsFooter() {
+export default function DataViewsFooter( {
+	className,
```
I've added a comment about this change in the PR description, just fyi.

youknowriad reviewed (commented):
I like the reasoning about separating pagination and footer but isn't the solution to instead expose `DataViewsPagination` rather `DataViewsFooter`.

Also `BulkActionsFooter`

youknowriad commented on packages/dataviews/CHANGELOG.automattic.md:
```diff
@@ -2,6 +2,12 @@
+- Expose self-sufficient DataViews subcomponents:
+	- `<DataViews.Search />`
+	- `<DataViews.Filters />`
+	- `<DataViews.ViewConfig />`
```
It feels like `ViewConfig` should be just `DataviewsViewConfigDropdown`

and `ViewTypeMenu` should be exposed separately rather than exporting the component wrapping both.

retrofox replied:
Good one. In fact, I suffered from that limitation in the story

retrofox replied:
According to https://github.com/WordPress/gutenberg/issues/63646, should we rename `ViewTypeMenu` to `LayoutSwitcher `?

retrofox replied:
renamed here https://github.com/Automattic/wp-calypso/pull/103070/commits/829ede1cf0298d9efaba401b4c73a696f3e0d32a. I like `LayoutSwitcher` more :-)

retrofox committed:
Export DataviewsViewConfigDropdown and update subcomponent reference

Ensures DataViews.ViewConfig exposes only the dropdown for clearer API and separation

retrofox committed:
Export ViewTypeMenu as a DataViews subcomponent

Adds ViewTypeMenu to DataViews for improved API flexibility

retrofox committed:
Update DataViews story to use new subcomponents

Showcases usage of Pagination, ViewConfig, and ViewTypeMenu together

oandregal commented on packages/dataviews/src/components/dataviews/stories/index.story.tsx:
```diff
@@ -130,44 +132,71 @@ function PlanetOverview( { planets }: { planets: SpaceObject[] } ) {
+			<DataViews.Filters />
+
+			<DataViews.Layout />
+			<DataViews.Pagination />
```
I'd remove this, it looks like a bug given it's also displayed in the header :)

oandregal commented on packages/dataviews/src/components/dataviews/stories/index.story.tsx:
```diff
@@ -130,44 +132,71 @@ function PlanetOverview( { planets }: { planets: SpaceObject[] } ) {
+				</HStack>
+			</Grid>
+
+			<DataViews.Filters />
```
Filters aren't displayed as far as I can see.

retrofox replied:
Correct, this is due to a lack of the Toggle button. https://github.com/Automattic/wp-calypso/pull/103050

oandregal replied:
I'd remove them from this PR and add them in that one. As a reviewer, it's helpful for me to review things that work without having to know that there's a related/future PR that fixes this one.

oandregal commented on packages/dataviews/src/components/dataviews/stories/index.story.tsx:
```diff
@@ -208,7 +237,7 @@ export const FreeComposition = () => {
+			fields={ fields }
 			onChangeView={ setView }
 			defaultLayouts={ defaultLayouts }
 		>
```
I still think passing planet as a prop to PlanetOverview is confusing. I know it's possible, but the point of the story is to highlight how to work with DataViews subcomponents, so I'd rather remove this property and access the filtered data via context within the component to highlight how that works.

retrofox replied:
It’s not possible. The story needs to reflect a real example, and that means the context isn’t accessible outside the API. [more](https://github.com/Automattic/wp-calypso/pull/103050#issuecomment-2846820614)

retrofox committed:
remove second Pagination instance from story

oandregal commented on packages/dataviews/src/components/dataviews/stories/index.story.tsx:
```diff
@@ -130,44 +132,71 @@ function PlanetOverview( { planets }: { planets: SpaceObject[] } ) {
+				<HStack justify="end">
+					<DataViews.Pagination />
+					<DataViews.ViewConfig />
+					<DataViews.ViewTypeMenu defaultLayouts={ defaultLayouts } />
```
`defaultLayouts` is a global. If this component were defined in its own file, it won't work.

retrofox replied:
I've exposed the `defaultLayouts` through the Context. What do you think

retrofox committed:
update changelog

oandregal commented on packages/dataviews/CHANGELOG.automattic.md:
```diff
@@ -2,6 +2,12 @@
 ## Next 
 
+- Expose self-sufficient DataViews subcomponents:
+	- `<DataViews.Search />`
```
Wasn't Search already exposed in a previous PR? Also, we could just add the new components to the existing list item below.

retrofox replied:
yes, updating.

oandregal commented on packages/dataviews/src/components/dataviews/stories/index.story.tsx:
```diff
@@ -130,44 +132,70 @@ function PlanetOverview( { planets }: { planets: SpaceObject[] } ) {
+				</HStack>
+
+				<HStack justify="end">
+					<DataViews.Pagination />
```
Riad mentioned this already, and I don't want it to go unnoticed: this component is a "footer" that includes pagination plus controls for bulkactions. They should be separate things.

retrofox replied:
Addressing. ~Sorry I missed that comment 🙁 Where did you leave it?~ Thanks 🙇‍♂️

retrofox committed:
Add defaultLayouts to DataViewsContext type and default value

Enables context consumers to access layout configuration directly

retrofox committed:
Refactor ViewTypeMenu and ViewConfig to use context for layouts

Removes prop drilling by accessing defaultLayouts directly from context

retrofox committed:
update story

retrofox committed:
remove unneded defaultLayouts prop

retrofox committed:
update changelog

retrofox committed:
pass a defaultLayouts in FreeComposition story

retrofox committed:
handle exposing Filters in follow ups

retrofox committed:
Export DataViewsPagination as Pagination subcomponent

Replaces DataViewsFooter with DataViewsPagination for accurate naming and usage

oandregal commented on packages/dataviews/src/components/dataviews/stories/index.story.tsx:
```diff
@@ -130,44 +132,68 @@ function PlanetOverview( { planets }: { planets: SpaceObject[] } ) {
+
+				<HStack justify="end">
+					<DataViews.Pagination />
+					<DataViews.ViewConfig />
```
The grid layout doesn't display the PreviewSizePicker component:

| This PR | Default story |
| --- | --- |
| <img width="370" alt="Screenshot 2025-05-02 at 13 46 56" src="https://github.com/user-attachments/assets/b7268740-2c3e-4c62-a91d-540fc541e8f8" /> | <img width="370" alt="Screenshot 2025-05-02 at 13 47 55" src="https://github.com/user-attachments/assets/32708e42-a82d-492c-ba43-368fa4e96c9c" /> |

retrofox replied:
fixed https://github.com/Automattic/wp-calypso/pull/103070/commits/1fedbc27c6da750d58e82a4d4aa48dcb8525a788. It was mentioned by @ntsekouras [here](https://github.com/Automattic/wp-calypso/pull/102996#discussion_r2071200052). 

I've created a follow-up WOOA7S-128

retrofox committed:
Export BulkActionsFooter as BulkActionToolbar subcomponent

retrofox committed:
Update stories to include BulkActionToolbar in DataViews examples

Demonstrates usage of bulk actions alongside pagination and view controls

retrofox committed:
changelog

retrofox committed:
wrapDataViews children and UI in dataviews-wrapper

retrofox committed:
rename component to LayoutSwitcher

ntsekouras commented on packages/dataviews/src/components/dataviews/index.tsx:
```diff
@@ -171,18 +175,31 @@ function DataViews< Item >( {
 			} }
 		>
-			{ children || defaultUI }
+			<div className="dataviews-wrapper" ref={ containerRef }>
```
I think this should a wrapper of `DataViewsLayout`, because since we free composition we could have a let's say split view of controls and the actual layout (sidebar).

Before it was fine as a top wrapper, since the layout of all sub components was vertical with full width.

retrofox replied:
Could we do it in a follow-up? I'd like to avoid increasing the chances here.

youknowriad replied:
This is also the only thing that I'm unsure about at the moment. I'm actually not entirely sure what's the right solution is.

retrofox replied:
follow issue WOOA7S-128

retrofox commented:
<!-- linear:isThreadRoot -->

Could we do it in a follow-up? I'd like to avoid increasing the chances here.

retrofox committed:
clean story

retrofox committed:
sort dataviews subcomponents by name

youknowriad commented on packages/dataviews/CHANGELOG.automattic.md:
```diff
@@ -2,6 +2,12 @@
+	- `<DataViews.LayoutSwitcher />`
+	- `<DataViews.Pagination />`
+	- `<DataViews.ViewConfig />`
 - Expose `<DataViews.Search />` component.
```
Can we merge these two changelog items?

retrofox committed:
tidy changelog
